### PR TITLE
C# Selenium Webdriver Sample

### PR DESCRIPTION
**Pull request by AhmedAbdoOrtiga**
#### Description of changes

- Implementing a sample C# project that uses Selenium.Axe 1.1.0.
- More samples will be added once we release a new "Selenium.Axe" Nuget that has the latest changes:
   - With Tags.
   - Disable Rules.

#### Pull request checklist

- [ ] If this PR addresses an existing issue, it is linked: Fixes #0000
- [ ] `yarn test` passes in all affected samples
- [ ] Added any applicable tests

### CODE REVIEW

File: CSharpSeleniumWebdriverSample/CSharpSeleniumWebdriverSample/tests/SampleTest.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+            string elementSelector1 = "ul";
+            int expectedNumberOfViolation1 = 2;
+            AxeResult results1 = RunScanOnGivenElementBySelector(elementSelector1);
+            results1.Violations.Should().HaveCount(expectedNumberOfViolation1);
+
+            string elementSelector2 = "ol";
+            int expectedNumberOfViolation2 = 1;
+            AxeResult results2 = RunScanOnGivenElementBySelector(elementSelector2);
+            results2.Violations.Should().HaveCount(expectedNumberOfViolation2);
```
Comment (devaradhanm) [inline CSharpSeleniumWebdriverSample/CSharpSeleniumWebdriverSample/tests/SampleTest.cs]: remove the validation on second element, as it doesnt provide any other additional usage that first validation provides

File: CSharpSeleniumWebdriverSample/CSharpSeleniumWebdriverSample/tests/SampleTest.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            string integrationTestTargetFile = Path.GetFullPath(samplePageURL);
+            string integrationTestTargetUrl = new Uri(integrationTestTargetFile).AbsoluteUri;
+            
+            this.InitDriver(browser);
+            LoadTestPage(integrationTestTargetUrl);
+
+            string elementSelector1 = "ul";
+            int expectedNumberOfViolation1 = 2;
+            AxeResult results1 = RunScanOnGivenElementBySelector(elementSelector1);
+            results1.Violations.Should().HaveCount(expectedNumberOfViolation1);
```
Comment (devaradhanm) [inline CSharpSeleniumWebdriverSample/CSharpSeleniumWebdriverSample/tests/SampleTest.cs]: may be just say  results1.Violations.Should().BeEmpty() ?

File: CSharpSeleniumWebdriverSample/CSharpSeleniumWebdriverSample/tests/SampleTest.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+
+        [TestMethod]
+        [DataRow("Chrome")]
+        [DataRow("Firefox")]
+        public void RunScanOnGivenElement(string browser)
+        {
+            string samplePageURL = @"src\samplePage.html";
+            string integrationTestTargetFile = Path.GetFullPath(samplePageURL);
+            string integrationTestTargetUrl = new Uri(integrationTestTargetFile).AbsoluteUri;
```
Comment (devaradhanm) [inline CSharpSeleniumWebdriverSample/CSharpSeleniumWebdriverSample/tests/SampleTest.cs]: this information is not relevant for this test. move this to LoadTestPage. & just call LoadTestPage("samplePage.html") or LoadTestPage()

File: CSharpSeleniumWebdriverSample/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+[![Screenshot of "dotnet test" build logs in sample build](./assets/screenshot-logs-tab.png)](https://dev.azure.com/accessibility-insights/axe-pipelines-samples/_build/results?buildId=2312&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9)
+
+The test pass/fail results display in the Tests tab of the build logs:
+
+[![Screenshot of Tests tab in sample build](./assets/screenshot-tests-tab.png)](https://dev.azure.com/accessibility-insights/axe-pipelines-samples/_build/results?buildId=2312&view=ms.vss-test-web.build-test-results-tab&runId=6462&resultId=100000&paneView=debug)
+
+## See it in action on your local machine
+
+1. Install the stable version of [Chrome](https://www.google.com/chrome/)
+1. Install the stable version of [Firefox] (https://www.mozilla.org/en-US/firefox/)
```
Comment (dbjorge) [inline CSharpSeleniumWebdriverSample/README.md]: Extra space between ] and ( breaking markdown formatting

File: CSharpSeleniumWebdriverSample/README.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+1. Clone this sample repository
+
+   ```sh
+   git clone https://github.com/microsoft/axe-pipelines-samples
+   ```
+
+1. Build the Test Project
```
Comment (dbjorge) [inline CSharpSeleniumWebdriverSample/README.md]: `dotnet test` implies the necessary parts of restore and build; you can just say `dotnet test` after the `cd` line, or just run `dotnet test ./csharp-selenium-webdriver-sample/CSharpSeleniumWebdriverSample.csproj` and omit the `cd` as well

File: CSharpSeleniumWebdriverSample/README.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+   dotnet build --no-restore -c $(buildConfiguration) -p:Version=$(version)
+   ```
+
+1. Run the tests!
+
+   ```sh
+   dotnet test --no-restore
+   ```
+
+   ![Screenshot of jest command showing all tests passing](./assets/screenshot-dotnet-tes-success.png)
```
Comment (dbjorge) [inline CSharpSeleniumWebdriverSample/README.md]: typo in filename? (`dotnet-tes-success`)

Comment (dbjorge) [inline CSharpSeleniumWebdriverSample/README.md]: Alt text should refer to "dotnet test" rather than jest

File: CSharpSeleniumWebdriverSample/README.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+The individual files in the sample contain comments that explain the important parts of each file in context.
+
+Some good places to start reading are:
+
+* [tests/SampleTest.cs](./tests/SampleTest.cs): C# test file that opens [src/samplePage.html](./src/samplePage.html) in a browser with Selenium and runs accessibility scans against it with Selenium.WebDriver.ChromeDriver and Selenium.WebDriver.GeckoDriver
+* [azure-pipelines.yml](./azure-pipelines.yml): Azure Pipelines config file that sets up our Continuous Integration and Pull Request builds
+
+## Tools and libraries used
+
+* [Microsoft.NET.Test.Sdk](https://github.com/microsoft/vstest/) to enable running tests, collect diagnostics data and report results.
```
Comment (dbjorge) [inline CSharpSeleniumWebdriverSample/README.md]: We can probably slim down this list a bit; I think we can combine the first two bullets into just ".NET Core" and the third bullet into just "MSTest"

File: CSharpSeleniumWebdriverSample/README.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+* [Microsoft.NETCore.App](https://dotnet.microsoft.com) a set of .NET API's that are included in the default .NET Core application model.
+* [MSTest.TestAdapter and MSTest.TestFramework](https://github.com/microsoft/testfx) MSTest V2 adapter to discover and execute MSTest Framework based tests.
+* [Selenium.Axe](https://github.com/TroyWalshProf/SeleniumAxeDotnet) to run an axe accessibility scan on the page from the Selenium browser.
+* [Selenium.WebDriver.ChromeDriver](https://github.com/jsakamoto/nupkg-selenium-webdriver-chromedriver/) to enable Selenium to drive Chrome.
```
Comment (dbjorge) [inline CSharpSeleniumWebdriverSample/README.md]: We should probably also include "Selenium.WebDriver" itself here (maybe adding an explicit PackageReference for it in the csproj too)

Comment (AhmedAbdoOrtiga) [reply CSharpSeleniumWebdriverSample/README.md]: We don't use it in the project, do we still need it?

File: CSharpSeleniumWebdriverSample/README.md (hunk 7 above)
Comment (AhmedAbdoOrtiga) [reply CSharpSeleniumWebdriverSample/README.md]: They first 2 bullets have different project URL, the third one is supposed to be 2 packages with the same project URL, that's why I merged them in one bullet

File: README.md [hunk 9]
```diff
@@ -9,6 +9,9 @@ The following sample projects specify the main technologies used. A team that us
 
 * **[Sample 1: typescript-selenium-webdriver-sample](./typescript-selenium-webdriver-sample)**: 
   * Useful for teams using Selenium and TypeScript/JavaScript. 
+  
+* **[Sample 2: CSharpSeleniumWebdriverSample](./c%23-selenium-webdriver-sample)**:
```
Comment (dbjorge) [inline README.md]: I'd recommend avoiding the # in the folder name; it makes it easy to form links that involve accidental anchor tags. Probably `csharp-selenium-webdriver-sample`

File: c#-selenium-webdriver-sample/azure-pipelines.yml [hunk 10]
```diff
@@ -0,0 +1,27 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+pool:
+    vmImage: windows-2019
+
+variables:
+  buildConfiguration: 'Release'
```
Comment (dbjorge) [inline c#-selenium-webdriver-sample/azure-pipelines.yml]: It doesn't look like `buildConfiguration` is used anymore

File: c#-selenium-webdriver-sample/WebdriverFactory.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+{
+    public class WebdriverFactory
+    {
+        public static IWebDriver GetWebdriver(BrowserType browser)
+        {
+            IWebdriver webdriver = null;
+
+            switch (browser)
+            {
+                //In case using Chrome Web Driver
```
Comment (dbjorge) [inline c#-selenium-webdriver-sample/WebdriverFactory.cs]: I think the comments above these case blocks can be omitted; they aren't adding any new information except what the case statement already says

File: c#-selenium-webdriver-sample/Webdriver.cs [hunk 12]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using OpenQA.Selenium;
+
+namespace CSharpSeleniumWebdriverSample
+{
+    public interface IWebdriver
```
Comment (dbjorge) [inline c#-selenium-webdriver-sample/Webdriver.cs]: I was pretty confused by this interface differing only in 1 character of casing with the real `IWebDriver` interface. Why is this necessary, rather than just returning `IWebDriver`s directly from a WebDriverFactory?

File: c#-selenium-webdriver-sample/SampleTest.cs [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+            string integrationTestTargetUrl = new Uri(integrationTestTargetFile).AbsoluteUri;
+
+            _webDriver.Navigate().GoToUrl(integrationTestTargetUrl);
+            _wait.Until(drv => drv.FindElement(By.TagName("main")));
+        }
+
+        private void InitDriver(BrowserType browser)
+        {
+            _webDriver = WebdriverFactory.GetWebdriver(browser);
+            _wait = new WebDriverWait(_webDriver, TimeSpan.FromSeconds(20));
```
Comment (dbjorge) [inline c#-selenium-webdriver-sample/SampleTest.cs]: I think we could make the tests a little simpler by using `_webDriver.Manage().Timeouts().ImplicitWait` in `InitDriver` and omitting the `WebDriverWait` usage throughout

File: c#-selenium-webdriver-sample/SampleTest.cs [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        public void RunScanOnGivenElement(BrowserType browser)
+        {
+            this.InitDriver(browser);
+            LoadTestPage();
+
+            AxeResult results1 = RunScanOnGivenElementBySelector("ul");
+            Assert.AreEqual(3, results1.Violations.Length);
+        }
+
+        private AxeResult RunScanOnGivenElementBySelector(string elementSelector)
```
Comment (dbjorge) [inline c#-selenium-webdriver-sample/SampleTest.cs]: I think it might be a little simpler to read if you inlined this helper (especially once the `_wait.Until` is removed).

File: c#-selenium-webdriver-sample/SampleTest.cs [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+        [TestMethod]
+        [DataRow(BrowserType.Chrome)]
+        [DataRow(BrowserType.Firefox)]
+        public void RunScanOnGivenElement(BrowserType browser)
+        {
+            this.InitDriver(browser);
+            LoadTestPage();
+
+            AxeResult results1 = RunScanOnGivenElementBySelector("ul");
```
Comment (dbjorge) [inline c#-selenium-webdriver-sample/SampleTest.cs]: Why results**1**?

File: c#-selenium-webdriver-sample/Webdriver.cs (hunk 12 above)
Comment (dbjorge) [reply c#-selenium-webdriver-sample/Webdriver.cs]: (I don't think the individual ChromeWebdriver or FirefoxWebdriver wrappers are necessary, either; I think the extra complexity of splitting the initialization across several files detracts from their use in a sample. Frankly, even the separate `WebdriverFactory` seems like a bit much for a sample; I think a helper method in the actual test class would be fine.

File: CSharpSeleniumWebdriverSample/README.md (hunk 8 above)
Comment (dbjorge) [reply CSharpSeleniumWebdriverSample/README.md]: We do use it in the project (ctrl+F for `using OpenQA.Selenium;`) - we just happen to be getting it indirectly via the `Selenium.Axe` dependency. But I think it's an important enough piece of the sample that we should note it explicitly in this list

File: c#-selenium-webdriver-sample/README.md [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+1. Install the stable version of [Chrome](https://www.google.com/chrome/)
+1. Install the stable version of [Firefox](https://www.mozilla.org/en-US/firefox/)
+1. Clone this sample repository
+
+   ```sh
+   git clone https://github.com/microsoft/axe-pipelines-samples
+   ```
+
+1. Build the Test Project
```
Comment (dbjorge) [inline c#-selenium-webdriver-sample/README.md]: "Run the tests"?

File: CSharpSeleniumWebdriverSample/README.md (hunk 7 above)
Comment (dbjorge) [reply CSharpSeleniumWebdriverSample/README.md]: I think for the purposes of the readme, it's more helpful to use project names as they would be recognized by a typical user, not necessarily just enumerating all the NuGet package names we happen to depend on. In particular, I think saying that we use ".NET Core" and "MSTest" adequately imply all of the other dependencies in the first 3 bullet points, but would be more concise and more likely to be recognized offhand by a consumer of the sample

File: c#-selenium-webdriver-sample/Webdriver.cs (hunk 12 above)
Comment (AhmedAbdoOrtiga) [reply c#-selenium-webdriver-sample/Webdriver.cs]: this was the first approach that we discussed to support multiple browsers, plus I don't think it's that complex, it's a known pattern that should be easy to follow, still we can change it if you believe a helper function would be better.

Comment (dbjorge) [reply c#-selenium-webdriver-sample/Webdriver.cs]: You can support multiple browsers with the DataRows of a BrowserType enum (this is as far as I remember us discussing) without using a doubly indirected FactoryFactory pattern :)

What we saw in user studies was that, for this sort of sample code, there is a significant number of users who will try to consume the sample by clicking around basically at random between files/links, and will rapidly abandon reading as soon as they start feeling confused or overwhelmed. Because of that, I think it's important to minimize the number of helper files in the sample; it's really important that the structure of the project lead people towards the important entry points.

I'm okay with keeping the separate `WebDriverFactory` static class if you feel it's more clear than making it a helper method in the test class, but I really think the `IWebDriverCreator`, `FirefoxWebDriverCreator`, and `ChromeWebDriverCreator` are adding an extra layer of abstraction for no benefit; if you were to inline the browser creators into the `WebDriverFactory`, it would actually make the factory **shorter**:

```cs
public static IWebDriver GetWebDriver(BrowserType browser)
{
    switch (browser)
    {
        case BrowserType.Chrome:
            var chromeDriverDirectory = Environment.GetEnvironmentVariable("ChromeWebDriver") ?? Environment.CurrentDirectory;
            return new ChromeDriver(chromeDriverDirectory);

        case BrowserType.Firefox:
            var geckoDriverDirectory = Environment.GetEnvironmentVariable("GeckoWebDriver") ?? Environment.CurrentDirectory;
            return new FirefoxDriver(geckoDriverDirectory);
    }
}
```

File: csharp-selenium-webdriver-sample/WebDriverFactory.cs [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+                    var chromeDriverDirectory = Environment.GetEnvironmentVariable("ChromeWebDriver") ?? Environment.CurrentDirectory;
+                    return new ChromeDriver(chromeDriverDirectory);
+
+                case BrowserType.Firefox:
+                    // Check if Gecko web driver eniroment varibale already defined so that Firefox version and the webdriver are the same, other wise use Selenium.WebDriver.GeckoDriver to indtall it
+                    var geckoDriverDirectory = Environment.GetEnvironmentVariable("GeckoWebDriver") ?? Environment.CurrentDirectory;
+                    return new FirefoxDriver(geckoDriverDirectory);
+            }
+
+            return null;
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/WebDriverFactory.cs]: Maybe `throw new ArgumentException($"Unknown browser type {browser}", nameof(browser));` would be more descriptive?

File: csharp-selenium-webdriver-sample/WebDriverFactory.cs [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    public class WebDriverFactory
+    {
+        public static IWebDriver GetWebDriver(BrowserType browser)
+        {   
+            switch (browser)
+            {
+                case BrowserType.Chrome:
+                    // Check if Chrome web driver eniroment varibale already defined so that Chrome version and the webdriver are the same, other wise use Selenium.WebDriver.ChromeDriver to indtall it
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/WebDriverFactory.cs]: Some misspellings in this comment; I'm okay with omitting it until the separate PR where we add the more verbose comments throughout
